src/charm.py

```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm for GitHub Actions Exporter on kubernetes."""
+
+import logging
+from typing import Dict
+
+from charms.prometheus_k8s.v0.prometheus_scrape import MetricsEndpointProvider
+from charms.traefik_k8s.v1.ingress import IngressPerAppRequirer
+from ops.charm import CharmBase, HookEvent, WorkloadEvent
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+
+from charm_state import CharmState
+from github_actions_exporter import exporter_environment, exporter_version, is_configuration_valid
+
+logger = logging.getLogger(__name__)
+
+GH_EXPORTER_CONTAINER_NAME = "github-actions-exporter"
+GH_EXPORTER_METRICS_PORT = 9101
+GH_EXPORTER_WEBHOOK_PORT = 8065
+
+
+class GithubActionsExporterOperatorCharm(CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args) -> None:
+        """Construct."""
+        super().__init__(*args)
+        self.framework.observe(
+            self.on.github_actions_exporter_pebble_ready,
+            self._on_github_actions_exporter_pebble_ready,
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.state: CharmState = CharmState(self)
+        self.ingress = IngressPerAppRequirer(
+            self,
+            port=GH_EXPORTER_WEBHOOK_PORT,
+            # We're forced to use the app's service endpoint
+            # as the ingress per app interface currently always routes to the leader.
+            # https://github.com/canonical/traefik-k8s-operator/issues/159
+            host=f"{self.app.name}-endpoints.{self.model.name}.svc.cluster.local",
+            strip_prefix=True,
+        )
+        self._metrics_endpoint = MetricsEndpointProvider(
+            self,
+            jobs=[
+                {
+                    "static_configs": [
+                        {
+                            "targets": [
+                                f"*:{GH_EXPORTER_METRICS_PORT}",
+                            ]
+                        }
+                    ]
+                }
+            ],
+        )
+
+    def _on_github_actions_exporter_pebble_ready(self, event: WorkloadEvent):
+        """Define and start a workload using the Pebble API.
+
+        Args:
+            event: Event triggering after pebble is ready.
+        """
+        container = event.workload
+        self.unit.status = MaintenanceStatus(f"Adding {container.name} layer to pebble")
+        container.add_layer(container.name, self._pebble_layer, combine=True)
+        container.replan()
+        self.unit.status = ActiveStatus()
+        version = exporter_version(container, self.state)
```

Comment:
@jdkandersson this is what I meant - if you rename functions and directly import them, it would read like `version = version(...)` and to me a bit of context is lost.